components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/impl/parsing/expressions/PostgresVaultNamedQueryExpressionParser.kt

```diff
@@ -0,0 +1,164 @@
+package net.corda.ledger.persistence.query.impl.parsing.expressions
+
+import net.corda.ledger.persistence.query.impl.parsing.And
+import net.corda.ledger.persistence.query.impl.parsing.As
+import net.corda.ledger.persistence.query.impl.parsing.Equals
+import net.corda.ledger.persistence.query.impl.parsing.From
+import net.corda.ledger.persistence.query.impl.parsing.GreaterThan
+import net.corda.ledger.persistence.query.impl.parsing.GreaterThanEquals
+import net.corda.ledger.persistence.query.impl.parsing.In
+import net.corda.ledger.persistence.query.impl.parsing.IsNotNull
+import net.corda.ledger.persistence.query.impl.parsing.IsNull
+import net.corda.ledger.persistence.query.impl.parsing.JsonArrayOrObjectAsText
+import net.corda.ledger.persistence.query.impl.parsing.JsonCast
+import net.corda.ledger.persistence.query.impl.parsing.JsonKeyExists
+import net.corda.ledger.persistence.query.impl.parsing.Keyword
+import net.corda.ledger.persistence.query.impl.parsing.LeftParentheses
+import net.corda.ledger.persistence.query.impl.parsing.LessThan
+import net.corda.ledger.persistence.query.impl.parsing.LessThanEquals
+import net.corda.ledger.persistence.query.impl.parsing.Like
+import net.corda.ledger.persistence.query.impl.parsing.NotEquals
+import net.corda.ledger.persistence.query.impl.parsing.Number
+import net.corda.ledger.persistence.query.impl.parsing.Or
+import net.corda.ledger.persistence.query.impl.parsing.Parameter
+import net.corda.ledger.persistence.query.impl.parsing.ParameterEnd
+import net.corda.ledger.persistence.query.impl.parsing.PathReference
+import net.corda.ledger.persistence.query.impl.parsing.PathReferenceWithSpaces
+import net.corda.ledger.persistence.query.impl.parsing.RightParentheses
+import net.corda.ledger.persistence.query.impl.parsing.Select
+import net.corda.ledger.persistence.query.impl.parsing.Token
+import net.corda.ledger.persistence.query.impl.parsing.Where
+
+class PostgresVaultNamedQueryExpressionParser : VaultNamedQueryExpressionParser {
+    private val stringPattern = Regex(
+        """(?<str>('[^']*)'|("[^"]*)")"""
+    )
+
+    @Suppress("MaxLineLength")
+    private val pathPattern = Regex(
+        """(?<path>(\$?[a-zA-Z_][a-zA-Z0-9_]*(\[([0-9]+|"[a-zA-Z_][a-zA-Z0-9_]*")])?)(\.[a-zA-Z_][a-zA-Z0-9_]*(\[([0-9]+|"[a-zA-Z_][a-zA-Z0-9_]*")])?)*)"""
+    )
+
+    private val numberPattern = Regex(
+        """(?<num>[0-9]+(\.[0-9]+)?([eE]-?[0-9]+)?)"""
+    )
+
+    @Suppress("MaxLineLength")
+    private val opsPattern = Regex(
+        """(?<op>(->>)|[+-/*=?]|<(=)?|>(=)?|==|!(=)?|(?i)\bas\b|(?i)\bfrom\b|(?i)\bselect\b|(?i)\bwhere\b|(?i)\band\b|(?i)\bor\b|(?i)\bis null\b|(?i)\bis not null\b|(?i)\bin\b|(?i)\blike\b)"""
+    )
+
+    private val jsonCastPattern = Regex("""(?<cast>::.*?)((->>)|[+*=]|&&|\|\||<(=)?|>(=)?|==|!(=)?|\s)""")
```

Comment:
```suggestion
    private val jsonCastPattern = Regex("""::(?<cast>.*?)((->>)|[+*=]|&&|\|\||<(=)?|>(=)?|==|!(=)?|\s)""")
```
That way you get the colon removed.

Reply:
I believe if you start your regexes with `^` they will only match at the given start position, so you can drop the index == startPosition.

Reply:
> I believe if you start your regexes with ^ they will only match at the given start position, so you can drop the index == startPosition.

I couldn't get this working myself.